src/backend/api/meals.js

```diff
@@ -3,12 +3,68 @@ const router = express.Router();
 const knex = require('../database');
 
 router.get('/', async (req, res) => {
+  let meals = knex.select('meal.*').from('meal');
+
+  if ('maxPrice' in req.query) {
+    meals = meals.whereBetween('price', [0, req.query.maxPrice]);
+  }
+
+  if ('availableReservations' in req.query) {
+    meals = meals
+      .join('reservation', 'reservation.meal_id', '=', 'meal.id')
+      .groupBy('meal.id');
+    if (req.query.availableReservations === 'true') {
+      meals = meals.having(
+        'meal.max_reservations',
+        '>',
+        knex.raw('SUM(reservation.number_of_guests)')
+      );
+    } else {
+      meals = meals.having(
+        'meal.max_reservations',
+        '=',
+        knex.raw('SUM(reservation.number_of_guests)')
+      );
+    }
+  }
+
+  if ('title' in req.query) {
+    meals = meals.whereRaw('LOWER(title) LIKE ?', [`%${req.query.title}%`]);
+  }
+
+  if ('dateAfter' in req.query) {
+    meals = meals.whereRaw('`when` > ?', [req.query.dateAfter]);
+  }
+
+  if ('dateBefore' in req.query) {
+    meals = meals.whereRaw('`when` < ?', [req.query.dateBefore]);
+  }
+
+  if ('limit' in req.query) {
+    meals = meals.limit(req.query.limit);
+  }
+
+  if (
+    'sort_key' in req.query &&
+    ['when', 'max_reservations', 'price'].includes(req.query.sort_key)
+  ) {
+    if (
+      'sort_dir' in req.query &&
+      ['asc', 'desc'].includes(req.query.sort_dir)
+    ) {
+      meals = meals.orderBy(req.query.sort_key, req.query.sort_dir);
+    } else {
+      meals = meals.orderBy(req.query.sort_key);
+    }
+  }
+
+  console.log('SQL', meals.toSQL().sql);
+
   try {
-    // knex syntax for selecting things. Look up the documentation for knex for further info
-    const meals = await knex.select().from('meal');
+    const mealsResponse = await meals;
 
-    if (meals.length !== 0) {
-      res.json(meals);
+    if (mealsResponse.length !== 0) {
+      res.json(mealsResponse);
     } else {
       res.status(404).json({ error: 'No meals found' });
```

Comment:
I would remove this `if` and just return the (potentially empty) list of meals. 404 responses make more sense if you're trying to get a single resource by ID but it doesn't exist.